File: c++/minimum-window-substring.cc
class Solution {
public:
    string minWindow(string s, string t) {
        int len = s.size();
        if (t.size() > len) return "";

        unordered_map<char, int> target;
        set<char> unique;

        for (char c : t) {
            target[c]++;
            unique.insert(c);
        }

        int left = 0;
        int right = 0;
        int matched = 0;

        int ansLeft = 0;
        int ansLen = 0;

        while (right < len && !unique.contains(s[right])) {
            right++;
        }
        if (right >= len) return "";
        left = right;

        for (; right < len; right++) {
            if (!unique.contains(s[right])) continue;
            if (--target[s[right]] == 0) matched++;

            if (s[right] == s[left]) {
                while (!unique.contains(s[left]) || target[s[left]] < 0 && left < right) {
                    if (unique.contains(s[left])) target[s[left]]++;
                    left++;
                }
            }   
            if (matched == unique.size() && !ansLen || ansLen > right - left + 1) {
                ansLen = right - left + 1;
                ansLeft = left;
            }
        }
        if (ansLen) return s.substr(ansLeft, ansLen);
        return "";
    }
};

File: python/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""

        needed = len(t)
        acquired = 0
        diffMap = Counter(t)
        windowStart = 0
        windowLen = 0

        left = 0
        for right in range(len(s)):
            curElem = s[right]
            if curElem in diffMap:
                if diffMap[curElem] > 0:
                    acquired += 1
                diffMap[curElem] -= 1
            if acquired < needed:
                continue

            while acquired == needed:
                if not windowLen or windowLen > right - left + 1:
                    windowStart = left
                    windowLen = right - left + 1
                if s[left] in diffMap:
                    diffMap[s[left]] += 1
                    if diffMap[s[left]] > 0:
                        acquired -= 1
                left += 1

        return s[windowStart:windowStart + windowLen] if windowLen > 0 else ""
